### ISSUE TITLE
Any Plan on openresty-openssl-1.1.1?

### ISSUE THREAD

**Issue by yongjianchn** [openresty/openresty#456]
Now the openresty-openssl is based on openssl-1.1.0 without TLS1.3 support.

**Comment by agentzh** [comment]
Yeah, we're gonna upgrade to 1.1.1 soon after upgrading our own openssl patches for this new version.

**Comment by rhuddleston** [comment]
Any estimated timeline for this effort?

**Comment by agentzh** [comment]
@spacewander may be able to give a rough estimate :)

**Comment by sparvu** [comment]
is this scoped for OpenResty 1.15.8.1 ? I hope you can release soon 1.15.8.1.

**Comment by agentzh** [comment]
@sparvu No, OpenSSL 1.1.1 support is decoupled from OpenResty releases. They are separate unless compatibility fixes are required in the OpenResty core.

**Comment by spacewander** [comment]
@rhuddleston 
I will try to support OpenSSL 1.1.1 in this month.

**Comment by vadi2** [comment]
I'm getting this, is the issue related to lack of 1.1.1 support?

```
2019/04/10 11:38:29 [notice] 2709#2709: start worker process 2710
2019/04/10 11:38:36 [error] 2710#2710: *1 lua entry thread aborted: runtime error: error loading module '_openssl.hmac' from file '/usr/local/lib/lua/5.1/_openssl.so':
	/usr/local/openresty/openssl/lib/libssl.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/local/lib/lua/5.1/_openssl.so)
stack traceback:
coroutine 0:
	[C]: in function 'require'
	content_by_lua(nginx.conf.compiled:22):2: in function <content_by_lua(nginx.conf.compiled:22):1>, client: 127.0.0.1, server: , request: "GET / HTTP/1.1", host: "localhost:8080"
```

**Comment by agentzh** [comment]
@vadi2 Your Lua module is linked against openssl 1.1.1 while your openresty is linked to an earlier version (presumably 1.1.0). Don't do that.

**Comment by vadi2** [comment]
I don't think I did anything; I installed openresty from the provided apt repository

**Comment by agentzh** [comment]
@vadi2 Your `/usr/local/lib/lua/5.1/_openssl.so` library should be compiled against OpenResty's OpenSSL's header files (from the `openresty-openssl-dev` deb package) and be linked against OpenResty's OpenSSL's library files (from the `openresty-openssl` package) at runtime.

**Comment by agentzh** [comment]
@vadi2 You cannot mix stuff compiled and linked against incompatible versions of OpenSSL. It's that simple.

**Comment by rhuddleston** [comment]
Any update on getting these patches merged?

**Comment by agentzh** [comment]
Superseded by PR #1503. We'll try merging #1503 after OpenResty 1.15.8.1 formal is out.

**Comment by thibaultcha** [comment]
Support for OpenSSL 1.1.1 in `ssl_session_fetch_by_lua*` is now merged across the board (patches, lua-nginx-module and CI test suites), thanks to @spacewander's work!

To use OpenSSL 1.1.1 with `ssl_session_fetch_by_lua*` today (that is, before the next formal release of OpenResty), make sure to apply the NGINX core patches documented here:

https://openresty.org/en/nginx-ssl-patches.html

E.g.:

```
$ wget https://openresty.org/download/openresty-1.15.8.1.tar.gz
$ tar -xvf openresty-1.15.8.1.tar.gz
$ cd openresty-1.15.8.1
$ curl -s https://raw.githubusercontent.com/openresty/openresty/master/patches/nginx-1.15.8-ssl_cert_cb_yield.patch | patch -d bundle/nginx-1.15.8 -p1
patching file src/event/ngx_event_openssl.c
Hunk #1 succeeded at 1580 (offset 370 lines).
$ curl -s https://raw.githubusercontent.com/openresty/openresty/master/patches/nginx-1.15.8-ssl_sess_cb_yield.patch | patch -d bundle/nginx-1.15.8 -p1
patching file src/event/ngx_event_openssl.c
patching file src/event/ngx_event_openssl.h
$ ./configure [...]
$ make && make install
```

**Comment by thibaultcha** [comment]
Given that OpenSSL 1.1.1 does not require any specific patches anymore for those features (see https://openresty.org/en/openssl-patches.html for reference), I am not sure if openresty-openssl-1.1.1 should be planned? cc @agentzh 

This issue was closed by the PR merge, but if openresty-openssl-1.1.1 is to be worked on, we could reopen it.

**Comment by rhuddleston** [comment]
This is great, any idea on a timeline for a next formal release of openresty that will include this work?

**Comment by LHertz** [comment]
> Support for OpenSSL 1.1.1 in `ssl_session_fetch_by_lua*` is now merged across the board (patches, lua-nginx-module and CI test suites), thanks to @spacewander's work!
> 
> To use OpenSSL 1.1.1 with `ssl_session_fetch_by_lua*` today (that is, before the next formal release of OpenResty), make sure to apply the NGINX core patches documented here:
> 
> https://openresty.org/en/nginx-ssl-patches.html
> 
> E.g.:
> 
> ```
> $ wget https://openresty.org/download/openresty-1.15.8.1.tar.gz
> $ tar -xvf openresty-1.15.8.1.tar.gz
> $ cd openresty-1.15.8.1
> $ curl -s https://raw.githubusercontent.com/openresty/openresty/master/patches/nginx-1.15.8-ssl_cert_cb_yield.patch | patch -d bundle/nginx-1.15.8 -p1
> patching file src/event/ngx_event_openssl.c
> Hunk #1 succeeded at 1580 (offset 370 lines).
> $ curl -s https://raw.githubusercontent.com/openresty/openresty/master/patches/nginx-1.15.8-ssl_sess_cb_yield.patch | patch -d bundle/nginx-1.15.8 -p1
> patching file src/event/ngx_event_openssl.c
> patching file src/event/ngx_event_openssl.h
> $ ./configure [...]
> $ make && make install
> ```

I did this exactly, but still got the error:
```
[crit] 58804#0: *8 lua: cannot yield in sess get cb: missing async sess get cb support in OpenSSL while fetching SSL session by lua,
```
I checked the OpenResty version output:
```
# /usr/local/openresty/bin/openresty -V
nginx version: openresty/1.15.8.1
built by gcc 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)
built with OpenSSL 1.1.1c  28 May 2019
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt=-O2 --add-module=../ngx_devel_kit-0.3.1rc1 --add-module=../echo-nginx-module-0.61 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2 --add-module=../set-misc-nginx-module-0.32 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.08 --add-module=../srcache-nginx-module-0.31 --add-module=../ngx_lua-0.10.15 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.33 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.7 --add-module=../rds-json-nginx-module-0.15 --add-module=../rds-csv-nginx-module-0.09 --add-module=../ngx_stream_lua-0.0.7 --with-ld-opt='-Wl,-rpath,/usr/local/openresty/luajit/lib -L/usr/local/lib -Wl,-E' --with-stream --with-stream_ssl_module --with-stream_ssl_preread_module --with-http_ssl_module
```
Did this have anything to do with the fact that I installed openssl 1.1.1 by building from the source?

**Comment by thriqon** [comment]
Are there plans to release an official Debian package for `openresty-openssl` with version `1.1.1`? If yes, do you have a rough timeline?

**Comment by thriqon** [comment]
@thibaultcha, @agentzh : Any news here? Or any pointers on how one could contribute?

**Comment by HeWhoWas** [comment]
TLS 1.3 support is required in order to support Low Latency HLS - given that this issue is closed, is it no longer being addressed?

**Comment by agentzh** [comment]
Yeah, we'll have openresty-openssl111 binary packages in our official binary repos.

**Comment by agentzh** [comment]
Maybe in the next few weeks or so.

**Comment by agentzh** [comment]
It should be a separate issue for the binary packages since the openssl 1.1.1 support has already been merged into the mainline code base anyway.

**Comment by agentzh** [comment]
Sorry for the late reply. Been very busy recently.

**Comment by AndreiG6** [comment]
Sorry but I don't quite follow. Does the latest OpenResty version (1.15.8.2) available for CentOS 7 support TLS 1.3, or are we still waiting on further openresty-openssl changes?

**Comment by Nattefrost** [comment]
Same issue, using MX Linux 19 (debian 9 based). I followed the code above and compiling openresty from sources worked properly  it looks like. Hope it can get packaged as a .deb soon though.

**Comment by jsanford-toast** [comment]
seeing this issue again with openresty/openresty:1.15.8.2-7-bionic. i tried openresty/openresty:1.15.8.1-4-bionic and had no issue. Just wondering if possibly this problem was reintroduced in 1.15.8.2-7.

`2020/03/06 00:01:56 [error] 18#18: init_by_lua error: error loading module 'ssl.core' from file '/usr/local/openresty/luajit/lib/lua/5.1/ssl.so':
	/usr/local/openresty/openssl/lib/libssl.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/local/openresty/luajit/lib/lua/5.1/ssl.so)
stack traceback:
	[C]: at 0x7f2bac587ab0
	[C]: in function 'require'
	/usr/local/openresty/luajit/share/lua/5.1/ssl.lua:8: in main chunk
	[C]: in function 'require'
	/usr/local/openresty/luajit/share/lua/5.1/ssl/https.lua:10: in main chunk
	[C]: in function 'require'
	...openresty/lualib/toast/clients/http/luasocket-client.lua:5: in main chunk
	[C]: in function 'require'
	.../local/openresty/lualib/toast/routing/route-resolver.lua:210: in function 'init'
	init_by_lua:25: in main chunk
nginx: [error] init_by_lua error: error loading module 'ssl.core' from file '/usr/local/openresty/luajit/lib/lua/5.1/ssl.so':
	/usr/local/openresty/openssl/lib/libssl.so.1.1: version `OPENSSL_1_1_1' not found (required by /usr/local/openresty/luajit/lib/lua/5.1/ssl.so)`
